File: Java_client/bim-portal-enhanced/src/main/java/com/pb40/bimportal/auth/TokenManager.java
package com.pb40.bimportal.auth;

import com.bimportal.client.model.JWTTokenPublicDto;
import com.pb40.bimportal.config.BimPortalConfig;
import java.time.Duration;
import java.time.Instant;
import java.util.Base64;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Thread-safe JWT token manager for BIM Portal authentication.
 *
 * <p>Handles token storage, validation, expiration checking, and user UUID extraction.
 */
public class TokenManager {

  private static final Logger logger = LoggerFactory.getLogger(TokenManager.class);

  // Patterns for extracting claims from JWT payload
  private static final Pattern USER_ID_PATTERN =
      Pattern.compile("\"(?:sub|userId|user_id|id)\"\\s*:\\s*\"([^\"]+)\"");
  private static final Pattern EXP_PATTERN = Pattern.compile("\"exp\"\\s*:\\s*(\\d+)");

  private volatile String accessToken;
  private volatile String refreshToken;
  private volatile Instant tokenExpiresAt;
  private volatile UUID currentUserId;

  private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
  private final ReentrantReadWriteLock.ReadLock readLock = lock.readLock();
  private final ReentrantReadWriteLock.WriteLock writeLock = lock.writeLock();

  /**
   * Store tokens from authentication response and extract user information.
   *
   * @param tokenResponse JWT token response from API
   */
  public void storeTokens(JWTTokenPublicDto tokenResponse) {
    if (tokenResponse == null) {
      logger.warn("Cannot store null token response");
      return;
    }

    writeLock.lock();
    try {
      this.accessToken = tokenResponse.getToken();
      this.refreshToken = tokenResponse.getRefreshToken();

      // Extract user UUID from JWT token
      this.currentUserId = extractUserIdFromToken(tokenResponse.getToken());

      Instant now = Instant.now();

      // ALWAYS use the JWT exp claim - it's the reliable source
      this.tokenExpiresAt = parseTokenExpiration(tokenResponse.getToken());
      logger.info("Using JWT expiration. Token expires at: {}", tokenExpiresAt);

      // Log token validity duration
      Duration validityDuration = Duration.between(now, tokenExpiresAt);
      long secondsUntilExpiry = validityDuration.getSeconds();
      long millisUntilExpiry = validityDuration.toMillis();

      logger.debug(
          "Token valid for {} seconds ({} milliseconds)", secondsUntilExpiry, millisUntilExpiry);

      // Additional warning if token is about to expire
      if (secondsUntilExpiry < 50) {
        logger.warn(
            "Token has only {} seconds remaining - consider refreshing soon", secondsUntilExpiry);
      }

      // Log user information
      if (currentUserId != null) {
        logger.info("Extracted user UUID from token: {}", currentUserId);
      } else {
        logger.warn("Could not extract user UUID from JWT token");
      }

    } finally {
      writeLock.unlock();
    }
  }

  /**
   * Get the current access token if valid.
   *
   * @return Access token or null if not available or expired
   */
  public String getAccessToken() {
    readLock.lock();
    try {
      if (accessToken == null) {
        logger.debug("No access token available");
        return null;
      }

      if (isTokenExpiring()) {
        logger.debug("Access token is expiring soon or expired");
        return null;
      }

      return accessToken;
    } finally {
      readLock.unlock();
    }
  }

  /**
   * Get the current refresh token.
   *
   * @return Refresh token or null if not available
   */
  public String getRefreshToken() {
    readLock.lock();
    try {
      return refreshToken;
    } finally {
      readLock.unlock();
    }
  }

  /**
   * Get the current user UUID extracted from the JWT token.
   *
   * @return User UUID or empty if not available
   */
  public Optional<UUID> getCurrentUserId() {
    readLock.lock();
    try {
      return Optional.ofNullable(currentUserId);
    } finally {
      readLock.unlock();
    }
  }

  /**
   * Check if the current token is expiring soon.
   *
   * @return True if token will expire within the configured margin
   */
  public boolean isTokenExpiring() {
    readLock.lock();
    try {
      if (tokenExpiresAt == null) {
        logger.debug("No expiration time set - treating as expired");
        return true;
      }

      Instant now = Instant.now();
      Instant refreshThreshold = now.plus(BimPortalConfig.TOKEN_REFRESH_MARGIN);

      boolean expiring = tokenExpiresAt.isBefore(refreshThreshold);

      if (expiring) {
        long secondsUntilExpiry = tokenExpiresAt.getEpochSecond() - now.getEpochSecond();
        logger.debug(
            "Token expiring: {} seconds remaining (refresh margin: {} seconds)",
            secondsUntilExpiry,
            BimPortalConfig.TOKEN_REFRESH_MARGIN.getSeconds());
      }

      return expiring;
    } finally {
      readLock.unlock();
    }
  }

  /**
   * Check if tokens are available.
   *
   * @return True if access token exists
   */
  public boolean hasTokens() {
    readLock.lock();
    try {
      return accessToken != null;
    } finally {
      readLock.unlock();
    }
  }

  /** Clear all stored tokens and user information. */
  public void clearTokens() {
    writeLock.lock();
    try {
      this.accessToken = null;
      this.refreshToken = null;
      this.tokenExpiresAt = null;
      this.currentUserId = null;
      logger.info("All tokens and user information cleared");
    } finally {
      writeLock.unlock();
    }
  }

  /**
   * Extract user UUID from JWT token payload. Looks for common user ID claims: sub, userId,
   * user_id, id
   *
   * @param token JWT token
   * @return User UUID or null if not found or parsing fails
   */
  private UUID extractUserIdFromToken(String token) {
    if (token == null || token.trim().isEmpty()) {
      return null;
    }

    try {
      String[] parts = token.split("\\.");
      if (parts.length < 2) {
        logger.warn("Invalid JWT token format for user ID extraction");
        return null;
      }

      // Decode the payload (second part of JWT)
      String payload = new String(Base64.getUrlDecoder().decode(parts[1]));
      logger.debug("JWT payload for user extraction: {}", payload);

      // Try to find user ID in common claim names
      Matcher matcher = USER_ID_PATTERN.matcher(payload);
      if (matcher.find()) {
        String userIdString = matcher.group(1);
        logger.debug("Found user ID string in JWT: {}", userIdString);

        try {
          // Try to parse as UUID
          UUID userId = UUID.fromString(userIdString);
          logger.info("Successfully extracted user UUID: {}", userId);
          return userId;
        } catch (IllegalArgumentException e) {
          logger.warn("User ID found in JWT is not a valid UUID: {}", userIdString);
          // Could return null or try other approaches
          return null;
        }
      } else {
        logger.warn("No user ID claim found in JWT payload. Checked for: sub, userId, user_id, id");

        // Log available claims for debugging
        logAvailableClaims(payload);
        return null;
      }

    } catch (Exception e) {
      logger.warn("Failed to extract user ID from JWT: {}", e.getMessage());
      return null;
    }
  }

  /**
   * Parse JWT token expiration from token payload. This is a fallback method when validTill is not
   * available in the response.
   *
   * @param token JWT token
   * @return Expiration instant or null if parsing fails
   */
  private Instant parseTokenExpiration(String token) {
    if (token == null || token.trim().isEmpty()) {
      return null;
    }

    try {
      String[] parts = token.split("\\.");
      if (parts.length < 2) {
        logger.warn("Invalid JWT token format");
        return null;
      }

      // Decode the payload (second part of JWT)
      String payload = new String(Base64.getUrlDecoder().decode(parts[1]));
      //            logger.debug("JWT payload: {}", payload);

      // Use pattern matching for more reliable extraction
      Matcher matcher = EXP_PATTERN.matcher(payload);
      if (matcher.find()) {
        String expValue = matcher.group(1);
        long expirationSeconds = Long.parseLong(expValue);

        Instant expiration = Instant.ofEpochSecond(expirationSeconds);
        logger.debug("Parsed token expiration from JWT exp claim: {}", expiration);

        return expiration;
      } else {
        logger.warn("No exp claim found in JWT token");
        return null;
      }

    } catch (Exception e) {
      logger.warn("Failed to parse token expiration from JWT: {}", e.getMessage());
      return null;
    }
  }

  /**
   * Log available claims in JWT payload for debugging purposes.
   *
   * @param payload JWT payload JSON string
   */
  private void logAvailableClaims(String payload) {
    try {
      // Simple extraction of claim names for debugging
      Pattern claimPattern = Pattern.compile("\"([^\"]+)\"\\s*:");
      Matcher matcher = claimPattern.matcher(payload);
      StringBuilder claims = new StringBuilder();
      while (matcher.find()) {
        if (claims.length() > 0) {
          claims.append(", ");
        }
        claims.append(matcher.group(1));
      }
      logger.debug("Available claims in JWT: {}", claims.toString());
    } catch (Exception e) {
      logger.debug("Could not extract claim names for debugging", e);
    }
  }

  /**
   * Get token status information for debugging.
   *
   * @return Status string
   */
  public String getTokenStatus() {
    readLock.lock();
    try {
      if (accessToken == null) {
        return "No token";
      }

      StringBuilder status = new StringBuilder();

      if (tokenExpiresAt == null) {
        status.append("Token available (expiration unknown)");
      } else {
        long secondsRemaining = tokenExpiresAt.getEpochSecond() - Instant.now().getEpochSecond();

        if (isTokenExpiring()) {
          status.append(
              String.format(
                  "Token expiring soon (expires in %d seconds at %s)",
                  secondsRemaining, tokenExpiresAt));
        } else {
          status.append(
              String.format(
                  "Token valid (expires in %d seconds at %s)", secondsRemaining, tokenExpiresAt));
        }
      }

      // Add user information
      if (currentUserId != null) {
        status.append(", User ID: ").append(currentUserId);
      } else {
        status.append(", User ID: not available");
      }

      return status.toString();
    } finally {
      readLock.unlock();
    }
  }

  /**
   * Manually set the token expiration time. This can be used when the actual expiration is known
   * from other sources.
   *
   * @param expiresAt The expiration instant
   */
  public void setTokenExpiration(Instant expiresAt) {
    writeLock.lock();
    try {
      this.tokenExpiresAt = expiresAt;
      logger.info("Token expiration manually set to: {}", expiresAt);
    } finally {
      writeLock.unlock();
    }
  }

  /**
   * Manually set the current user ID. This can be used if the user ID is obtained from other
   * sources.
   *
   * @param userId The user UUID
   */
  public void setCurrentUserId(UUID userId) {
    writeLock.lock();
    try {
      this.currentUserId = userId;
      logger.info("Current user ID manually set to: {}", userId);
    } finally {
      writeLock.unlock();
    }
  }
}

File: Python_client/auth/token_manager.py
import threading
from datetime import datetime, timezone
from typing import Optional, Union, Dict, Any
import jwt

# Import from your new Pydantic models instead of generated client
from models import JWTTokenPublicDto
from .auth_config import TOKEN_REFRESH_MARGIN, logger

class TokenManager:
    """
    Manages the storage and lifecycle of authentication tokens in a thread-safe manner.
    Adapted to work with Pydantic models.
    """

    def __init__(self):
        self._access_token: Optional[str] = None
        self._refresh_token: Optional[str] = None
        self._expires_at: Optional[datetime] = None
        self._lock = threading.Lock()

    def set_token(self, token_data: Union[JWTTokenPublicDto, Dict[str, Any]]) -> None:
        """
        Sets the access and refresh tokens from various response types.

        This method can handle:
        1. A JWTTokenPublicDto object from Pydantic models.
        2. A simple dictionary with a 'token' key (from the old client).
        3. A dictionary with 'token', 'refreshToken', and 'validTill' keys.
        """
        with self._lock:
            logger.debug("Setting new token in TokenManager.")
            if isinstance(token_data, JWTTokenPublicDto):
                self._access_token = token_data.token if token_data.token is not None else None
                self._refresh_token = token_data.refreshToken if token_data.refreshToken is not None else None
                # The DTO provides a datetime object directly
                self._expires_at = token_data.validTill if token_data.validTill is not None else None
            elif isinstance(token_data, dict):
                self._access_token = token_data.get("token")
                self._refresh_token = token_data.get("refreshToken")

                # Handle expiration
                if "validTill" in token_data and token_data["validTill"]:
                    # Parse ISO format string and ensure timezone-aware
                    expires_str = token_data["validTill"]
                    if expires_str.endswith('Z'):
                        expires_str = expires_str[:-1] + '+00:00'
                    self._expires_at = datetime.fromisoformat(expires_str)
                    # Ensure timezone-aware
                    if self._expires_at.tzinfo is None:
                        self._expires_at = self._expires_at.replace(tzinfo=timezone.utc)
                elif self._access_token:
                    # Fallback to decoding the JWT to get the 'exp' claim
                    try:
                        decoded_token = jwt.decode(self._access_token, options={"verify_signature": False})
                        exp = decoded_token.get("exp")
                        if exp:
                            self._expires_at = datetime.fromtimestamp(exp, tz=timezone.utc)
                    except jwt.PyJWTError as e:
                        logger.error(f"Failed to decode JWT to get expiration: {e}")
                        self._expires_at = None
            else:
                logger.error(f"Unsupported token data type: {type(token_data)}")
                self._clear_tokens()

            if self._access_token:
                 logger.debug(f"Token set successfully. Expiration: {self._expires_at}")
            else:
                 logger.warning("Attempted to set token, but no access token was found in the provided data.")

    def get_access_token(self) -> Optional[str]:
        """Returns the current access token."""
        with self._lock:
            return self._access_token

    def get_refresh_token(self) -> Optional[str]:
        """Returns the current refresh token."""
        with self._lock:
            return self._refresh_token

    def is_token_expiring(self) -> bool:
        """
        Checks if the access token is missing, expired, or about to expire.
        """
        with self._lock:
            if not self._access_token or not self._expires_at:
                logger.debug("Token is considered expiring because it's missing or has no expiration.")
                return True

            # Ensure timezone-aware comparison
            now = datetime.now(timezone.utc)
            
            # Ensure expires_at is timezone-aware
            if self._expires_at.tzinfo is None:
                self._expires_at = self._expires_at.replace(tzinfo=timezone.utc)

            is_expiring = self._expires_at <= (now + TOKEN_REFRESH_MARGIN)
            if is_expiring:
                logger.debug(f"Token is expiring. Expiration: {self._expires_at}, Now: {now}")
            return is_expiring

    def clear_tokens(self) -> None:
        """Clears all stored token data."""
        with self._lock:
            logger.debug("Clearing all tokens.")
            self._access_token = None
            self._refresh_token = None
            self._expires_at = None

    def _clear_tokens(self) -> None:
        """Internal method to clear tokens (called by set_token on error)."""
        self.clear_tokens()